### PR TITLE
Speciation

### PR DESCRIPTION
**Pull request by portaloffreedom**
Implement generic speciation in Revolve (inspired by NEAT)

### CODE REVIEW

File: pyrevolve/gazebo/connect.py [hunk 1]
```diff
@@ -254,7 +254,7 @@ async def do_request(self, msg):
         Performs a request. The only requirement
         of `msg` is that it has an `id` attribute.
 
-        Publishing of the request is always yielded to prevent multiple messages
+        Publishing of the srequest is always yielded to prevent multiple messages
```
Comment (DaanZ) [inline pyrevolve/gazebo/connect.py]: typo

File: pyrevolve/genotype/plasticoding/decoder.py [hunk 2]
```diff
@@ -124,8 +124,9 @@ def __init__(self, id_, conf, command_sentence):
         self.edges = {}
 
     def decode_sentence(self):
-        self.body._id = self._id if type(self._id) == str and self._id.startswith('robot') \
-            else f'robot_{self._id}'
+        # self.body._id = self._id if type(self._id) == str and self._id.startswith('robot') \
+        #     else f'robot_{self._id}'
```
Comment (DaanZ) [inline pyrevolve/genotype/plasticoding/decoder.py]: Is there a reason to keep this commented or is there a reason to remove the commented code.

File: pyrevolve/evolution/population/population.py [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+        data_path = self.config.experiment_management.experiment_folder
+        for r, d, f in os.walk(os.path.join(data_path, f'selectedpop_{gen_num}')):
+            for file in f:
+                if 'body' in file:
+                    _id = file.split('.')[0].split('_')[-2]+'_'+file.split('.')[0].split('_')[-1]
+                    self.individuals.append(
+                        self.config.experiment_management.load_individual(_id, self.config))
+
+    def load_offspring(self,
+                             last_snapshot: int,
```
Comment (DaanZ) [inline pyrevolve/evolution/population/population.py]: Indent does not align with self argument.

File: pyrevolve/evolution/speciation/species_collection.py [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+class SpeciesIterator(Iterator):
+    """
+    Concrete Iterators implement various traversal algorithms. These classes
+    store the current traversal position at all times.
+    """
+
+    """
+    `_position` stores the current traversal position.
+    """
+    _position: int = None
```
Comment (DaanZ) [inline pyrevolve/evolution/speciation/species_collection.py]: Is there a reason for the argument duplication?

File: pyrevolve/util/supervisor/analyzer_queue.py [hunk 5]
```diff
@@ -11,6 +11,7 @@ class AnalyzerQueue(SimulatorQueue):
 
     def __init__(self, n_cores: int, settings, port_start=11345, simulator_cmd='gzserver'):
         super(AnalyzerQueue, self).__init__(n_cores, settings, port_start, simulator_cmd)
+        self._enable_play_pause = False
```
Comment (DaanZ) [inline pyrevolve/util/supervisor/analyzer_queue.py:14]: Not used in analyzer queue code, so why is it added here?

File: pyrevolve/util/supervisor/simulator_queue.py [hunk 6]
```diff
@@ -123,6 +124,9 @@ async def _worker_evaluate_robot(self, connection, robot, future, conf):
     async def _simulator_queue_worker(self, i):
         try:
             self._free_simulator[i] = True
+            if self._enable_play_pause:
```
Comment (DaanZ) [inline pyrevolve/util/supervisor/simulator_queue.py:127]: Why is this addition necessary?

File: pyrevolve/util/supervisor/simulator_queue.py [hunk 7]
```diff
@@ -24,6 +24,7 @@ def __init__(self, n_cores: int, settings, port_start=11345, simulator_cmd=None)
         self._robot_queue = asyncio.Queue()
         self._free_simulator = [True for _ in range(n_cores)]
         self._workers = []
+        self._enable_play_pause = True
```
Comment (DaanZ) [inline pyrevolve/util/supervisor/simulator_queue.py:27]: Do other people benefit from pausing and resetting the simulator?

**Review by DaanZ — APPROVED** [review]
Please review the small changes, but nothing ground-breaking that I found by quickly looking through the code. 
I would like to refactor some of the larger files in species and population, but I will keep this for the refactor branch. This should also include more standardized tests, fortunately you already started for the speciation code.

File: pyrevolve/util/supervisor/analyzer_queue.py (hunk 5 above)
Comment (portaloffreedom) [reply pyrevolve/util/supervisor/analyzer_queue.py:14]: Because the simulator queue can pause the simulator when not running robots, but the analyzer cannot.
(the `AnalyzerQueue` is superclassing the `SimulatorQueue`)
